src/bech32.cpp

```diff
@@ -227,4 +431,103 @@ DecodeResult Decode(const std::string& str) {
     return {result, std::move(hrp), data(values.begin(), values.end() - 6)};
 }
 
+/** Find index of an incorrect character in a Bech32 string. */
+std::string LocateErrors(const std::string& str, std::vector<int>& error_locations) {
+    if (str.size() > 90) {
+        push_range(90, str.size(), error_locations);
+        return "Bech32 string too long";
+    }
+    if (!CheckCharacters(str, error_locations)){
+        return "Invalid character or mixed case";
+    }
+    size_t pos = str.rfind('1');
+    if (pos == str.npos) {
+        return "Missing separator";
+    }
+    if (pos == 0 || pos + 7 > str.size()) {
+        error_locations.push_back(pos);
+        return "Invalid separator position";
+    }
+    std::string hrp;
+    for (size_t i = 0; i < pos; ++i) {
+        hrp += LowerCase(str[i]);
+    }
+
+    size_t length = str.size() - 1 - pos; // length of data part
+    data values(length);
+    for (size_t i = pos + 1; i < str.size(); ++i) {
+        unsigned char c = str[i];
+        int8_t rev = CHARSET_REV[c];
+        if (rev == -1) {
+            error_locations.push_back(i);
+            return "Invalid Base 32 character";
+        }
+        values[i - pos - 1] = rev;
+    }
+
+    // We attempt error detection with both bech32 and bech32m, and choose the one with the fewest errors
+    // We can't simply use the segwit version, because that may be one of the errors
+    for (Encoding encoding : {Encoding::BECH32, Encoding::BECH32M}) {
+        std::vector<int> possible_errors;
+        uint32_t residue = PolyMod(Cat(ExpandHRP(hrp), values)) ^ EncodingConstant(encoding);
+        if (residue != 0) {
+            uint32_t syn = Syndrome(residue);
+            int s0 = syn & 0x3FF;
+            int s1 = (syn >> 10) & 0x3FF;
+            int s2 = syn >> 20;
+            int l_s0 = GF1024_LOG[s0];
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/16807#discussion_r733208374

> @ryanofsky Alternatively, would it help if there was an (additional) `& 0x3FF` inside the `[]`?

I think most ideal thing would be for it to use something like [`std::array::at`](https://en.cppreference.com/w/cpp/container/array/at) and throw an exception in case of a bug. Alternative of adding `& 0x3FF` to lookups could help avoid undefined behavior, but also clutter the code and mask potential bugs. What's best just depends on what your priorities are. No strong opinions from me!